java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestDoExchange.java

```diff
@@ -0,0 +1,407 @@
+      iv.allocateNew();
+
+      stream.getWriter().start(root);
+      for (int i = 0; i < 10; i++) {
+        iv.setSafe(0, i);
+        root.setRowCount(1);
+        if (i % 2 == 0) {
+          stream.getWriter().putNext();
+        } else {
+          buf = allocator.buffer(4);
+          buf.writeInt(i);
+          stream.getWriter().putNext(buf);
+        }
+
+        Assert.assertTrue(reader.next());
+        Assert.assertTrue(reader.hasRoot());
+        final ArrowBuf metadata = reader.getLatestMetadata();
+        if (i % 2 == 0) {
+          Assert.assertNull(metadata);
+        } else {
+          Assert.assertNotNull(metadata);
+          Assert.assertEquals(i, metadata.getInt(0));
+        }
+        Assert.assertEquals(root.getSchema(), reader.getSchema());
+        Assert.assertEquals(i, ((IntVector) reader.getRoot().getVector("a")).get(0));
+      }
+
+      buf = allocator.buffer(4);
+      buf.writeInt(126);
+      stream.getWriter().putMetadata(buf);
+      Assert.assertTrue(reader.next());
+      Assert.assertNotNull(reader.getLatestMetadata());
+      Assert.assertEquals(126, reader.getLatestMetadata().getInt(0));
+
+      stream.getWriter().completed();
+
+      // Ensure the stream is drained. Else, we race with the server for shutdown: we'll go and shut down the
+      // server/allocator before it can finish cleanup.
+      Assert.assertFalse(reader.next());
+    }
+  }
+
+  /** Write some data, have it transformed, then read it back. */
+  @Test
+  public void testTransform() throws Exception {
+    final Schema schema = new Schema(Arrays.asList(
+        Field.nullable("a", new ArrowType.Int(32, true)),
+        Field.nullable("b", new ArrowType.Int(32, true))));
+    try (final FlightClient.ExchangeReaderWriter stream =
+             client.doExchange(FlightDescriptor.command(EXCHANGE_TRANSFORM))) {
+      // Write data to the stream
+      final FlightStream reader = stream.getReader();
+      final FlightClient.ClientStreamListener writer = stream.getWriter();
+      try (final VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator)) {
+        writer.start(root);
+        for (int batchIndex = 0; batchIndex < 10; batchIndex++) {
+          for (final FieldVector rawVec : root.getFieldVectors()) {
+            final IntVector vec = (IntVector) rawVec;
+            for (int row = 0; row < batchIndex; row++) {
+              vec.setSafe(row, row);
+            }
+          }
+          root.setRowCount(batchIndex);
+          writer.putNext();
+        }
+      }
+      // Indicate that we're done writing so that the server does not expect more data.
+      writer.completed();
+
+      // Start reading back data.
+      Assert.assertEquals(schema, reader.getSchema());
+      final VectorSchemaRoot root = reader.getRoot();
+      for (int batchIndex = 0; batchIndex < 10; batchIndex++) {
+        Assert.assertTrue("We got back batch " + batchIndex, reader.next());
```

Comment:
should this read "No batch received" (is boolean actually communicated on failure?)